### PR TITLE
batches: select changesets, publication state from preview

### PR DESCRIPTION
**Pull request by courier-new**
Part 1 of #23383. Portions of this lifted from or inspired by #22658.

**Note:** This PR is temporarily based on #23960 since it copies the same pattern fixed there.

This PR includes:
- some light cleanup/followup to #23821
- selecting the preview changesets
- selecting the desired publish status from a dropdown

image

This PR does not include:
- recalculating the actions on the backend when a publish status modification is selected
- applying the new publish statuses when you click "Apply"
- disabling the "Apply" button until a publish status modification is selected

(Those will each be a follow-up PR, but this one I think is the chunkiest)

This absolutely duplicates a lot of code between `BatchChangeChangesets` and `PreviewList`, but I was not feeling particularly motivated to refactor it because
1. we're trying to get this out sooner rather than later
2. it's only duplicated across 2 places, not 3 or 4 or 5 or 10
3. we're slowly moving towards getting rid of the preview page anyway

### CODE REVIEW

File: client/shared/dev/generateGraphQlOperations.js [hunk 1]
```diff
@@ -71,7 +71,7 @@ async function generateGraphQlOperations() {
           JSONValue: 'unknown',
           GitObjectID: 'string',
           JSONCString: 'string',
-          PublishedValue: "boolean | 'draft'",
+          PublishedValue: "boolean | 'draft' | null",
```
Comment (eseliger) [inline client/shared/dev/generateGraphQlOperations.js:74]: I don't think we should do that. In GraphQL, nullability is specified in the schema  by omitting the `!`, making it `Nullable<T>`. Where is this change needed? Can we fixup the schema instead?

File: client/web/src/enterprise/batches/preview/list/backend.ts [hunk 2]
```diff
@@ -325,3 +327,33 @@ export const queryChangesetSpecFileDiffs = ({
...
     )
+
+export const queryPublishableChangesetSpecIDs = ({
+    batchSpec,
+    search,
+    currentState,
+    action,
+}: BatchSpecApplyPreviewVariables): Observable<Scalars['ID'][]> => {
+    const request = (after: string | null): Observable<BatchSpecApplyPreviewConnectionFields> =>
+        queryChangesetApplyPreview({
```
Comment (eseliger) [inline client/web/src/enterprise/batches/preview/list/backend.ts:338]: Can we add a separate method for this? This loads way more data than required here. This endpoint is relatively heavy in compute.

File: client/web/src/enterprise/batches/preview/utils.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        node.targets.__typename !== 'VisibleApplyPreviewTargetsUpdate'
+    ) {
+        return null
+    }
+    if (node.targets.changesetSpec.description.__typename !== 'GitBranchChangesetDescription') {
+        return null
+    }
+    if (node.targets.changesetSpec.description.published !== null) {
+        return null
+    }
```
Comment (eseliger) [inline client/web/src/enterprise/batches/preview/utils.ts:27]: Just for my understanding: This guards against setting the publication state for already published changesets, right?

File: client/web/src/enterprise/batches/preview/utils.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * the applied changeset is visible to the user applying, will be attaching or updating
+ * the changeset, is not an existing reference, and has not had its publish status set
+ * from the batch spec file. Returns the id of the changeset spec if it is publishable
+ * from the UI, or null if for any reason it is not.
+ *
+ * @param node the `ChangesetApplyPreviewFields` node to check
+ */
+export const getPublishableChangesetSpecID = (
+    node: VisibleChangesetApplyPreviewFields | HiddenChangesetApplyPreviewFields
+): string | null => {
```
Comment (eseliger) [inline client/web/src/enterprise/batches/preview/utils.ts:15]:
```suggestion
): Scalars['ID'] | null => {
```

File: client/web/src/enterprise/batches/preview/list/backend.ts [hunk 5]
```diff
@@ -325,3 +327,33 @@ export const queryChangesetSpecFileDiffs = ({
...
+        })
+
+    return request(null).pipe(
+        expand(connection => (connection.pageInfo.hasNextPage ? request(connection.pageInfo.endCursor) : EMPTY)),
+        reduce(
+            (previous, next) =>
+                previous.concat(
+                    next.nodes
+                        .map(node => getPublishableChangesetSpecID(node))
+                        .filter((maybeID): maybeID is string => maybeID !== null)
```
Comment (eseliger) [inline client/web/src/enterprise/batches/preview/list/backend.ts:354]:
```suggestion
                        .filter((maybeID): maybeID is Scalars['ID'] => maybeID !== null)
```

_should_ enable you to remove the cast below :)

File: client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx [hunk 6]
```diff
@@ -194,6 +207,50 @@ export const VisibleChangesetApplyPreviewNode: React.FunctionComponent<VisibleCh
...
+    if (!changesetSpecID) {
+        return (
+            <div className="p-2">
+                <InputTooltip
+                    id="select-changeset-hidden"
+                    type="checkbox"
+                    className="btn"
+                    checked={false}
+                    disabled={true}
+                    tooltip="You cannot currently modify the publish status for this changeset"
```
Comment (eseliger) [inline client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx:234]: should we add a small entry to the FAQ "Why do I see You cannot currently modify the publish status for this changeset" explaning the logic behind it?

**Review by eseliger — APPROVED** [review]
Nice! Very neat that we can reuse the action dropdown from the changesets page :) 

One little thing I found while looking through chromatic: On xs screens, the checkbox seems so random, maybe we add a small label or don't break the line after it? 
image

File: client/shared/dev/generateGraphQlOperations.js (hunk 1 above)
Comment (courier-new) [reply client/shared/dev/generateGraphQlOperations.js:74]: You know I ran into this sometime over the weekend and honestly can't remember where it was that I was seeing the issue with `null`. Let me just remove it. 😛

File: client/web/src/enterprise/batches/preview/list/backend.ts (hunk 2 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/preview/list/backend.ts:338]: Yes, thank you, I was thinking about that too but forgot to leave myself a TODO comment. Will add that.

File: client/web/src/enterprise/batches/preview/utils.ts (hunk 3 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/preview/utils.ts:27]: Slightly more specifically, it should guard against setting the publication state *from the UI* for any changeset whose publication state is already being controlled *from the spec file*. So basically, spec file still takes precedence.

File: client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx (hunk 6 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx:234]: I think that's a great idea! I'll include that on #23384.

**Comment by courier-new** [comment]
> On xs screens, the checkbox seems so random, maybe we add a small label or don't break the line after it?

Thanks for mentioning this! I feel like a label would be slightly more helpful here since there's just so much going on on the screen when you get down to that size, so here's what I came up with:

image

**Comment by eseliger** [comment]
Nice 👍 Maybe not word break before status? 🤔

File: client/web/src/enterprise/batches/MultiSelectContext.tsx [hunk 7]
```diff
@@ -13,8 +14,8 @@ export type MultiSelectContextSelected = Set<string> | 'all'
 export interface MultiSelectContextState {
     // State fields. These must not be mutated other than through the mutator
     // functions below, but may be read at any time.
-    selected: MultiSelectContextSelected
-    visible: Set<string>
+    readonly selected: MultiSelectContextSelected
+    readonly visible: Set<string>
```
Comment (LawnGnome) [inline client/web/src/enterprise/batches/MultiSelectContext.tsx:18]: TIL. (Thanks for fixing that!)

File: client/web/src/enterprise/batches/MultiSelectContext.tsx [hunk 8]
```diff
@@ -42,13 +46,16 @@ const defaultState = (): MultiSelectContextState => ({
...
+    deselectAll: noop,
+    deselectVisible: noop,
+    deselectSingle: noop,
+    selectAll: noop,
+    selectVisible: noop,
+    toggleAll: noop,
+    toggleVisible: noop,
+    toggleSingle: noop,
+    selectSingle: noop,
+    setVisible: noop,
```
Comment (LawnGnome) [inline client/web/src/enterprise/batches/MultiSelectContext.tsx:58]: This is the most ridiculous request in the world, but can we make this alphabetical, please?

```suggestion
    selectAll: noop,
    selectSingle: noop,
    selectVisible: noop,
    toggleAll: noop,
    toggleSingle: noop,
    toggleVisible: noop,

    setVisible: noop,
```

File: client/web/src/enterprise/batches/MultiSelectContext.tsx [hunk 9]
```diff
@@ -1,3 +1,4 @@
+import { noop } from 'lodash'
```
Comment (LawnGnome) [inline client/web/src/enterprise/batches/MultiSelectContext.tsx:1]: This is a good import. :+1:

File: client/web/src/enterprise/batches/preview/list/PreviewListHeader.tsx [hunk 10]
```diff
@@ -1,12 +1,29 @@
...
-export const PreviewListHeader: React.FunctionComponent<PreviewListHeaderProps> = () => (
+export const PreviewListHeader: React.FunctionComponent<PreviewListHeaderProps> = ({
+    allSelected,
+    toggleSelectAll,
+}) => (
     <>
         <span className="p-2 d-none d-sm-block" />
+        {toggleSelectAll && (
+            <div className="d-flex p-2 align-items-center">
+                <input
```
Comment (LawnGnome) [inline client/web/src/enterprise/batches/preview/list/PreviewListHeader.tsx:16]: Should we make this an `InputTooltip`, even though it doesn't have the possibility of being disabled, for consistency?

File: client/web/src/enterprise/batches/preview/list/PreviewSelectRow.tsx [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+import { queryPublishableChangesetSpecIDs as _queryPublishableChangesetSpecIDs } from './backend'
+
+const ACTIONS: Action[] = [
+    {
+        type: 'unpublish',
+        buttonLabel: 'Unpublish',
+        dropdownTitle: 'Unpublish',
+        dropdownDescription:
+            'Do not publish selected changesets on the codehost. Note, we cannot unpublish a published changeset.',
```
Comment (LawnGnome) [inline client/web/src/enterprise/batches/preview/list/PreviewSelectRow.tsx:20]: Slight wording suggestion (but I defer to @rrhyne if he has thoughts):

```suggestion
            'Do not publish selected changesets on the codehost. Note that a changeset that has already been published cannot be unpublished.',
```

File: client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx [hunk 12]
```diff
@@ -194,6 +207,49 @@ export const VisibleChangesetApplyPreviewNode: React.FunctionComponent<VisibleCh
...
+        <InputTooltip
+            id="select-changeset-hidden"
+            type="checkbox"
+            className="btn"
+            checked={false}
+            disabled={true}
+            tooltip="You cannot currently modify the publish status for this changeset"
+        />
+    ) : (
+        <input
```
Comment (LawnGnome) [inline client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx:235]: Again, `InputTooltip` for consistency?

File: client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx [hunk 13]
```diff
@@ -194,6 +207,49 @@ export const VisibleChangesetApplyPreviewNode: React.FunctionComponent<VisibleCh
...
+    }, [selectable, changesetSpecID])
+
+    const input = !changesetSpecID ? (
+        <InputTooltip
+            id="select-changeset-hidden"
+            type="checkbox"
+            className="btn"
+            checked={false}
+            disabled={true}
+            tooltip="You cannot currently modify the publish status for this changeset"
```
Comment (LawnGnome) [inline client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx:232]: I'm not too fussed about this, but I've been tending towards "publication state" rather than "publish status" for wording. As long as whatever we use is consistent, though, I'm OK with it.

**Comment by courier-new** [comment]
Rebased this on `main` now that #23960 is merged.

**Review by LawnGnome — APPROVED** [review]
I share Erik's wish for a backend query that can more efficiently provide which changeset specs are publishable, but assuming that's dealt with in a follow up PR, I'm otherwise good with this, modulo a few really nitpicky things below. Nice work!

File: client/web/src/enterprise/batches/preview/list/backend.ts (hunk 2 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/preview/list/backend.ts]: So I initially interpreted this as "add a separate method on the _frontend_ to only query the data necessary", but after talking with @LawnGnome I realized that you may have actually meant "add a separate method on the _backend_ that returns the publishable changeset spec ids". Were you more thinking of that latter interpretation?

Regardless of which you meant, I did go ahead and [implement that frontend optimization](https://github.com/sourcegraph/sourcegraph/pull/23966/commits/b93488bf13f911b569be1baf26df9398b1fae4e1), which should lessen the load at least slightly. But I'm happy to give it another go with the backend store method approach instead, or file a GitHub issue to do so as a follow-up. 🙂

File: client/web/src/enterprise/batches/MultiSelectContext.tsx (hunk 8 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/MultiSelectContext.tsx:58]: Hahaha I LOVE alphabetizing object keys actually, I just didn't want to here since you initially set the ordering! It would be my pleasure. 😂

File: client/web/src/enterprise/batches/preview/list/PreviewListHeader.tsx (hunk 10 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/preview/list/PreviewListHeader.tsx:16]: Sure why not!

File: client/web/src/enterprise/batches/preview/list/backend.ts (hunk 2 above)
Comment (eseliger) [reply client/web/src/enterprise/batches/preview/list/backend.ts]: I meant frontend :) I think it should not be a frontend developers/api consumers concern to switch to a separate API just because it's faster and we're not capable of making the other one fast in the backend, too. So, if we were to optimize that query speed, I don't think it should incur a schema change.

Comment (courier-new) [reply client/web/src/enterprise/batches/preview/list/backend.ts]: Got it, all right, thanks so much for clarifying! 😄

File: client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx (hunk 13 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx:232]: I'm not too attached to "publish status" -- I focused on it since that was the wording used in the "Apply" banner from the designs, but I like how you're tending, too, and could go either way. @rrhyne do you want to weigh in?

File: client/web/src/enterprise/batches/preview/list/PreviewSelectRow.tsx (hunk 11 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/preview/list/PreviewSelectRow.tsx:20]: It's funny, I actually originally had almost literally this, but it does make the description extend onto a third line:



It's probably okay since the other two action descriptions are so short?

Comment (rrhyne) [reply client/web/src/enterprise/batches/preview/list/PreviewSelectRow.tsx]: Slightly shorter: 

Note: a changeset that has been published cannot be unpublished.

Comment (courier-new) [reply client/web/src/enterprise/batches/preview/list/PreviewSelectRow.tsx]: Much better:
image

File: client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx (hunk 13 above)
Comment (rrhyne) [reply client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx:232]: Ditto what courier-new said.

Comment (courier-new) [reply client/web/src/enterprise/batches/preview/list/VisibleChangesetApplyPreviewNode.tsx:232]: All right, let's go with "publication state" so we match the name of the API parameter.
